File: Back-end/app/models/session.py
import enum
import uuid
from datetime import datetime, timezone

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    String,
    Table,
    Text,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.base import Base


class SessionStatus(str, enum.Enum):
    NORMAL = "normal"
    CAMBIO_SALON = "cambio_salon"
    RETRASADO = "retrasado"


class SessionModality(str, enum.Enum):
    PRESENCIAL = "presencial"
    VIRTUAL = "virtual"
    HIBRIDO = "hibrido"


class SessionTrack(str, enum.Enum):
    IA = "ia"
    CIBERSEGURIDAD = "ciberseguridad"
    IOT = "iot"
    DESARROLLO = "desarrollo"
    DATOS = "datos"
    INNOVACION = "innovacion"


class SessionEventType(str, enum.Enum):
    CONFERENCE = "conference"
    WORKSHOP = "workshop"
    SYMPOSIUM = "symposium"
    PANEL = "panel"


session_registrations = Table(
    "session_registrations",
    Base.metadata,
    Column("user_id", UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), primary_key=True),
    Column("session_id", UUID(as_uuid=True), ForeignKey("sessions.id", ondelete="CASCADE"), primary_key=True),
    Column(
        "registered_at",
        DateTime(timezone=True),
        nullable=True,
        default=lambda: datetime.now(timezone.utc),
    ),
)


class Session(Base):
    __tablename__ = "sessions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    titulo = Column(String(500), nullable=False, index=True)
    descripcion = Column(Text, nullable=True)
    ponente = Column(String(255), nullable=False)
    afiliacion = Column(String(255), nullable=True)
    speaker_id = Column(String(50), nullable=True)
    track = Column(Enum(SessionTrack), nullable=False)
    event_type = Column(Enum(SessionEventType), nullable=False)
    dia = Column(String(10), nullable=False, index=True)
    hora_inicio = Column(String(5), nullable=False)
    hora_fin = Column(String(5), nullable=False)
    salon = Column(String(255), nullable=False)
    salon_anterior = Column(String(255), nullable=True)
    modalidad = Column(Enum(SessionModality), nullable=False)
    status_logistico = Column(Enum(SessionStatus), nullable=False, default=SessionStatus.NORMAL)
    link_virtual = Column(String(1000), nullable=True)
    link_verificado = Column(Boolean, nullable=False, default=False)
    cupos_totales = Column(Integer, nullable=False)
    inscritos = Column(Integer, nullable=False, default=0)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    timestamp_actualizacion = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )
    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    descripcion_ponente = Column(Text, nullable=True)
    foto_ponente_url = Column(String(1000), nullable=True)
    es_conferencista_principal = Column(Boolean, nullable=False, default=False)

    created_by_user = relationship("User", back_populates="sessions", foreign_keys=[created_by])
    registered_users = relationship(
        "User",
        secondary=session_registrations,
        back_populates="registered_sessions",
    )

    def __repr__(self) -> str:
        return f"<Session id={self.id} titulo={self.titulo!r}>"

File: Front-end/src/types/session.js
/**
 * ===================================
 * Tipos de Datos de Sesión — Contrato API
 * ===================================
 *
 * Este archivo define la estructura de un objeto Sesión y
 * todos los enums / constantes relacionados. Cuando el backend (FastAPI)
 * sea implementado, estos tipos servirán como contrato entre
 * frontend y backend.
 */

/** @enum {string} */
export const SESSION_STATUS = Object.freeze({
  NORMAL: 'Normal',
  CAMBIO_SALON: 'Cambio de Salón',
  RETRASADO: 'Retrasado',
});

/** @enum {string} */
export const SESSION_MODALITY = Object.freeze({
  PRESENCIAL: 'Presencial',
  VIRTUAL: 'Virtual',
  HIBRIDO: 'Híbrido',
});

/** @enum {string} */
export const SESSION_TRACK = Object.freeze({
  IA: 'Inteligencia Artificial',
  CIBERSEGURIDAD: 'Ciberseguridad',
  IOT: 'Internet de las Cosas',
  DESARROLLO: 'Desarrollo de Software',
  DATOS: 'Ciencia de Datos',
  INNOVACION: 'Innovación y Tendencias',
});

/** @enum {string} */
export const SESSION_EVENT_TYPE = Object.freeze({
  CONFERENCE: 'Conferencia',
  WORKSHOP: 'Taller',
  SYMPOSIUM: 'Simposio',
  PANEL: 'Panel'
});

/** @enum {string} */
export const SESSION_ROOMS = Object.freeze({
  AUDITORIO_CE4: 'Auditorio CE4',
  SALA_MULTIPLE: 'Sala Múltiple',
  AUDITORIO_PPAL: 'Auditorio Principal',
  SALA_VIRTUAL: 'Sala Virtual 1',
  SALA_VIRTUAL_2: 'Sala Virtual 2'
});

/**
 * @typedef {Object} Speaker
 * @property {string}      id          - Identificador único del ponente
 * @property {string}      name        - Nombre completo
 * @property {string}      role        - Cargo / Título (ej: "Ph.D. en Computación")
 * @property {string}      affiliation - Institución o empresa
 * @property {string}      bio         - Biografía corta
 * @property {string|null} photo_url   - URL de la foto de perfil
 * @property {string}      country     - País de origen
 */

/**
 * @typedef {Object} Session
 * @property {string}            id                      - Identificador único
 * @property {string}            titulo                  - Título de la sesión
 * @property {string}            ponente                 - Nombre del ponente
 * @property {string}            afiliacion              - Afiliación del ponente
 * @property {string}            track                   - Track o tema
 * @property {string}            event_type              - Valor de SESSION_EVENT_TYPE
 * @property {string}            speaker_id              - ID del ponente (referencia a Speaker.id)
 * @property {string}            dia                     - Día (cadena de fecha ISO)
 * @property {string}            hora_inicio             - Hora de inicio (HH:mm)
 * @property {string}            hora_fin                - Hora de fin (HH:mm)
 * @property {string}            salon                   - Nombre del salón
 * @property {string}            salon_anterior          - Salón anterior (si hubo cambio)
 * @property {string}            modalidad               - Valor de SESSION_MODALITY
 * @property {string}            status_logistico        - Valor de SESSION_STATUS
 * @property {boolean}           link_verificado         - Si el enlace virtual está verificado
 * @property {string|null}       link_virtual            - URL de la sesión virtual
 * @property {string}            timestamp_actualizacion - Marca de tiempo ISO 8601 de la última actualización
 * @property {string}            descripcion             - Descripción corta
 */